File: backend/app/services/inventory_service.py
from typing import Optional
from fastapi import HTTPException, status
from app.database import store, supabase_client

class InventoryService:
    @staticmethod
    def adjust_stock(product_id: str, delta: int, reason: Optional[str] = None) -> dict:
        """
        Adjust stock quantity by delta (positive or negative).
        """
        if supabase_client:
            try:
                res = supabase_client.table("products").select("*").eq("id", product_id).single().execute()
                if res.data:
                    current = res.data.get("stock_quantity", 0)
                    new_stock = max(0, current + delta)
                    upd = supabase_client.table("products").update({
                        "stock_quantity": new_stock
                    }).eq("id", product_id).execute()
                    return upd.data[0] if upd.data else res.data
            except Exception as e:
                print(f"Supabase stock adjust error: {e}")

        if product_id not in store.products:
            raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Product not found")

        prod = store.products[product_id]
        current = prod.get("stock_quantity", 0)
        new_stock = max(0, current + delta)
        prod["stock_quantity"] = new_stock
        return prod

inventory_service = InventoryService()

File: src/server/services/inventory.service.ts
import { store, supabaseClient, isProduction } from '../database';
import { Product } from '../types';

export class InventoryService {
  async adjustStock(productId: string, delta: number, reason?: string): Promise<Product> {
    if (supabaseClient) {
      const { data: currentProd, error: fetchErr } = await supabaseClient
        .from('products')
        .select('*')
        .eq('id', productId)
        .single();

      if (fetchErr || !currentProd) {
        if (isProduction) {
          const err = new Error(`Product '${productId}' not found.`) as any;
          err.status = 404;
          throw err;
        }
      } else {
        const current = currentProd.stock_quantity ?? 0;
        if (delta < 0 && current < Math.abs(delta)) {
          const err = new Error(
            `Insufficient stock for '${currentProd.name}'. Available: ${current}, Requested: ${Math.abs(delta)}.`
          ) as any;
          err.status = 400;
          throw err;
        }

        const newStock = Math.max(0, current + delta);
        let updateQuery = supabaseClient
          .from('products')
          .update({ stock_quantity: newStock })
          .eq('id', productId);

        // Atomic concurrency check: ensure stock has not been reduced concurrently
        if (delta < 0) {
          updateQuery = updateQuery.gte('stock_quantity', Math.abs(delta));
        }

        const { data: updated, error: updErr } = await updateQuery.select().single();

        if (updErr || !updated) {
          if (delta < 0) {
            const err = new Error(
              `Concurrent stock conflict or insufficient stock for product '${currentProd.name}'.`
            ) as any;
            err.status = 400;
            throw err;
          }
          if (isProduction) {
            const err = new Error(updErr?.message || 'Failed to update product stock.') as any;
            err.status = 400;
            throw err;
          }
        } else {
          // Keep local store in sync if present
          if (store.products[productId]) {
            store.products[productId].stock_quantity = newStock;
            store.products[productId].inventory_count = newStock;
          }
          return updated as Product;
        }
      }
    } else if (isProduction) {
      throw new Error('Supabase database client required for inventory operations in production.');
    }

    const prod = store.products[productId];
    if (!prod) {
      const err = new Error('Product not found') as any;
      err.status = 404;
      throw err;
    }

    const current = prod.stock_quantity ?? 0;
    if (delta < 0 && current < Math.abs(delta)) {
      const err = new Error(
        `Insufficient stock for '${prod.name || 'Product'}'. Available: ${current}, Requested: ${Math.abs(delta)}.`
      ) as any;
      err.status = 400;
      throw err;
    }

    const newStock = Math.max(0, current + delta);
    prod.stock_quantity = newStock;
    prod.inventory_count = newStock;
    return prod;
  }
}

export const inventoryService = new InventoryService();
